File: ios/Classes/search/SearchModels.m
//
// Created by Yohom Bao on 2018-12-16.
//

#import "SearchModels.h"
#import "AMapSearchObj.h"
#import "NSArray+Rx.h"
#import "NSString+GeoPoint.h"
#import "MJExtension.h"


//region RoutePlanParam
@implementation RoutePlanParam {

}

- (NSString *)description {
    NSMutableString *description = [NSMutableString stringWithFormat:@"<%@: ", NSStringFromClass([self class])];
    [description appendFormat:@"self.from=%@", self.from];
    [description appendFormat:@", self.to=%@", self.to];
    [description appendFormat:@", self.mode=%i", self.mode];
    [description appendFormat:@", self.passedByPoints=%@", self.passedByPoints];
    [description appendFormat:@", self.avoidPolygons=%@", self.avoidPolygons];
    [description appendFormat:@", self.avoidRoad=%p", self.avoidRoad];
    [description appendString:@">"];
    return description;
}

@end
//endregion


//region UnifiedDriveRouteResult
@implementation UnifiedDriveRouteResult {

}
- (instancetype)initWithAMapRouteSearchResponse:(AMapRouteSearchResponse *)response {
    if ([self init]) {
        _startPos = response.route.origin;
        _targetPos = response.route.destination;
        _taxiCost = response.route.taxiCost;
        _paths = [response.route.paths map:^(AMapPath *path) {
            return [[UnifiedDrivePathResult alloc] initWithAMapPath:path];
        }];
    }
    return self;
}

@end


@implementation UnifiedDrivePathResult {
}

- (instancetype)initWithAMapPath:(AMapPath *)path {
    if ([self init]) {
        _strategy = path.strategy;
        _tolls = path.tolls;
        _tollDistance = path.tollDistance;
        _totalTrafficlights = path.totalTrafficLights;
        _steps = [path.steps map:^(AMapStep *step) {
            return [[UnifiedDriveStepResult alloc] initWithAMapStep:step];
        }];
        _restriction = path.restriction;
    }
    return self;
}

@end

@implementation UnifiedDriveStepResult {

}
- (instancetype)initWithAMapStep:(AMapStep *)step {
    if ([self init]) {
        _instruction = step.instruction;
        _orientation = step.orientation;
        _road = step.road;
        _distance = step.distance;
        _duration = step.duration;
        _polyline = [step.polyline stringToAMapGeoPoint];
        _action = step.action;
        _assistantAction = step.assistantAction;
        _tolls = step.tolls;
        _tollDistance = step.tollDistance;
        _tollRoad = step.tollRoad;
        _routeSearchCityList = [step.cities map:^(AMapCity *city) {
            return [[UnifiedRouteSearchCityResult alloc] initWithAMapCity:city];
        }];
        _TMCs = [step.tmcs map:^(AMapTMC *tmc) {
            return [[UnifiedTMCResult alloc] initWithAMapTMC:tmc];
        }];
    }
    return self;
}

@end

@implementation UnifiedRouteSearchCityResult {

}
- (instancetype)initWithAMapCity:(AMapCity *)step {
    if ([self init]) {
        _districts = [step.districts map:^(AMapDistrict *district) {
            return [[UnifiedDistrictResult alloc] initWithAMapDistrict:district];
        }];
    }
    return self;
}

@end

@implementation UnifiedTMCResult {

}
- (instancetype)initWithAMapTMC:(AMapTMC *)tmc {
    if ([self init]) {
        _distance = tmc.distance;
        _status = tmc.status;
        _polyline = [tmc.polyline stringToAMapGeoPoint];
    }
    return self;
}

@end

@implementation UnifiedDistrictResult {

}
- (instancetype)initWithAMapDistrict:(AMapDistrict *)district {
    if ([self init]) {
        _districtName = district.name;
        _districtAdcode = district.adcode;
    }
    return self;
}

@end
//endregion


//region UnifiedGeocodeResult
@implementation UnifiedGeocodeResult {

}

- (instancetype)initWithAMapGeocodeSearchResponse:(AMapGeocodeSearchResponse *)response {
    self = [super init];
    if (self) {
        _geocodeAddressList = [response.geocodes map:^(AMapGeocode* geocode) {
            return [[UnifiedGeocodeAddress alloc] initWithAMapGeocode:geocode];
        }];
    }

    return self;
}

@end

@implementation UnifiedGeocodeAddress {

}
- (instancetype)initWithAMapGeocode:(AMapGeocode *)geocode {
    self = [super init];
    if (self) {
        _formatAddress = geocode.formattedAddress;
        _province = geocode.province;
        _city = geocode.city;
        _adcode = geocode.adcode;
        _district = geocode.district;
        _township = geocode.township;
        _neighborhood = geocode.neighborhood;
        _building = geocode.building;
        _level = geocode.level;
        _latLng = geocode.location;
    }

    return self;
}

@end

@implementation UnifiedGeocodeQuery {

}
@end
//endregion


//region UnifiedPoiResult
@implementation UnifiedPoiResult {

}
- (instancetype)initWithPoiResult:(AMapPOISearchResponse *)result {
    if ([super init]) {
        _pageCount = result.count;
        _pois = [result.pois map:^(AMapPOI *poi) {
            return [[PoiItem alloc] initWithAMapPOI:poi];
        }];
        _searchSuggestionCitys = [result.suggestion.cities map:^(AMapCity *amapCity) {
            return [[SuggestionCity alloc] initWithAMapSuggestion:amapCity];
        }];
        _searchSuggestionKeywords = result.suggestion.keywords;
    }
    return self;
}

@end

@implementation PoiItem {

}
- (instancetype)initWithAMapPOI:(AMapPOI *)aMapPOI {
    if ([super init]) {
        _poiId = aMapPOI.uid;
        _adCode = aMapPOI.adcode;
        _adName = aMapPOI.district;
        _businessArea = aMapPOI.businessArea;
        _cityCode = aMapPOI.citycode;
        _cityName = aMapPOI.city;
        _direction = aMapPOI.direction;
        _distance = aMapPOI.distance;
        _email = aMapPOI.email;
        _indoorData = [[IndoorData alloc] initWithAMapIndoorData:aMapPOI.indoorData];
        _isIndoorMap = aMapPOI.hasIndoorMap;
        _latLonPoint = aMapPOI.location;
        _enter = aMapPOI.enterLocation;
        _exit = aMapPOI.exitLocation;
        _parkingType = aMapPOI.parkingType;
        _photos = [aMapPOI.images map:^(AMapImage *image) {
            return [[Photo alloc] initWithAMapImage:image];
        }];
        _poiExtension = [[PoiExtension alloc] initWithAMapPOIExtension:aMapPOI.extensionInfo];
        _postcode = aMapPOI.postcode;
        _provinceCode = aMapPOI.pcode;
        _provinceName = aMapPOI.province;
        _shopID = aMapPOI.shopID;
        _snippet = aMapPOI.address;
        _subPois = [aMapPOI.subPOIs map:^(AMapSubPOI *subPOI) {
            return [[SubPoiItem alloc] initWithSubPoi:subPOI];
        }];
        _tel = aMapPOI.tel;
        _title = aMapPOI.name;
        _typeCode = aMapPOI.typecode;
        _typeDes = aMapPOI.type;
        _website = aMapPOI.website;
        _gridCode = aMapPOI.gridcode;
    }
    return self;
}

@end

@implementation SuggestionCity {

}
- (instancetype)initWithAMapSuggestion:(AMapCity *)suggestion {
    if ([super init]) {
        _cityName = suggestion.city;
        _cityCode = suggestion.citycode;
        _adCode = suggestion.adcode;
        _suggestionNum = suggestion.num;
        _districts = suggestion.districts;
    }
    return self;
}

@end

@implementation SubPoiItem {

}
- (instancetype)initWithSubPoi:(AMapSubPOI *)subPOI {
    if ([super init]) {
        _poiId = subPOI.uid;
        _title = subPOI.name;
        _subName = subPOI.sname;
        _distance = subPOI.distance;
        _latLonPoint = subPOI.location;
        _snippet = subPOI.address;
        _subTypeDes = subPOI.subtype;
    }
    return self;
}

@end

@implementation PoiExtension {

}
- (instancetype)initWithAMapPOIExtension:(AMapPOIExtension *)poiExtension {
    if ([super init]) {
        _opentime = poiExtension.openTime;
        _rating = [NSString stringWithFormat:@"%f", poiExtension.rating];
        _cost = poiExtension.cost;
    }
    return self;
}

@end

@implementation Photo {

}
- (instancetype)initWithAMapImage:(AMapImage *)image {
    if ([super init]) {
        _title = image.title;
        _url = image.url;
    }
    return self;
}

@end

@implementation IndoorData {

}
- (instancetype)initWithAMapIndoorData:(AMapIndoorData *)indoorData {
    if ([super init]) {
        _floor = indoorData.floor;
        _floorName = indoorData.floorName;
        _poiId = indoorData.pid;
    }
    return self;
}

@end
//endregion


//region UnifiedPoiSearchQuery
@implementation UnifiedPoiSearchQuery {

}
- (AMapPOIKeywordsSearchRequest *)toAMapPOIKeywordsSearchRequest {
    AMapPOIKeywordsSearchRequest *request = [[AMapPOIKeywordsSearchRequest alloc] init];
    request.types = _category;
    if (_distanceSort) {
        request.sortrule = 0;
    } else {
        request.sortrule = 1;
    }
    request.offset = _pageSize;
    request.page = _pageNum;
    request.building = _building;
    request.requireExtension = _requireExtension;
    request.requireSubPOIs = _requireSubPois;
    request.keywords = _query;
    request.city = _city;
    request.cityLimit = _cityLimit;
    request.location = _location;
    return request;
}

- (AMapPOIAroundSearchRequest *)toAMapPOIAroundSearchRequest {
    AMapPOIAroundSearchRequest *request = [[AMapPOIAroundSearchRequest alloc] init];
    request.types = _category;
    if (_distanceSort) {
        request.sortrule = 0;
    } else {
        request.sortrule = 1;
    }
    request.offset = _pageSize;
    request.page = _pageNum;
    request.building = _building;
    request.requireExtension = _requireExtension;
    request.requireSubPOIs = _requireSubPois;
    request.keywords = _query;
    request.city = _city;
    request.location = _location;
    request.radius = _searchBound.range;
    return request;
}

- (AMapPOIPolygonSearchRequest *)toAMapPOIPolygonSearchRequest {
    AMapPOIPolygonSearchRequest *request = [[AMapPOIPolygonSearchRequest alloc] init];
    request.types = _category;
    if (_distanceSort) {
        request.sortrule = 0;
    } else {
        request.sortrule = 1;
    }
    request.offset = _pageSize;
    request.page = _pageNum;
    request.building = _building;
    request.requireExtension = _requireExtension;
    request.requireSubPOIs = _requireSubPois;
    request.keywords = _query;

    NSArray <AMapGeoPoint *> *polygonList = [_searchBound.polyGonList map:^(NSDictionary *location) {
        return [AMapGeoPoint mj_objectWithKeyValues:location];
    }];
    request.polygon = [AMapGeoPolygon polygonWithPoints:polygonList];
    return request;
}


@end
//endregion


//region UnifiedRoutePoiSearchQuery
@implementation UnifiedRoutePoiSearchQuery {

}
- (AMapRoutePOISearchRequest *)toAMapRoutePOISearchRequestLine {
    AMapRoutePOISearchRequest *result = [[AMapRoutePOISearchRequest alloc] init];
    result.origin = _from;
    result.destination = _to;
    result.strategy = _mode;
    result.searchType = (AMapRoutePOISearchType) _searchType;
    result.range = _range;
    return result;
}

- (AMapRoutePOISearchRequest *)toAMapRoutePOISearchRequestPolygon {
    AMapRoutePOISearchRequest *result = [[AMapRoutePOISearchRequest alloc] init];
    result.origin = _from;
    result.destination = _to;
    result.strategy = _mode;
    result.searchType = (AMapRoutePOISearchType) _searchType;
    result.range = _range;
    result.polyline = [_polylines map:^(NSDictionary *location) {
        return [AMapGeoPoint mj_objectWithKeyValues:location];
    }];
    return result;
}
@end
//endregion


//region UnifiedRoutePOISearchResult
@implementation UnifiedRoutePOISearchResult {

}
- (instancetype)initWithAMapRoutePOISearchResponse:(AMapRoutePOISearchResponse *)result {
    if ([self init]) {
        _query = nil;
        _routePoiList = [result.pois map:^(AMapPOI *poi) {
            return [[UnifiedRoutePOIItem alloc] initWithAMapPOI:poi];
        }];
    }
    return self;
}

@end

@implementation UnifiedRoutePOIItem {

}
- (instancetype)initWithAMapPOI:(AMapPOI *)aMapPOI {
    if ([self init]) {
        _id = aMapPOI.uid;
        _title = aMapPOI.name;
        _point = aMapPOI.location;
        _distance = aMapPOI.distance;
        _duration = 0;
    }
    return self;
}

@end
//endregion


//region UnifiedSearchBound
@implementation UnifiedSearchBound {

}
@end
//endregion

File: android/src/main/kotlin/me/yohom/amapbase/search/SearchModels.kt
package me.yohom.amapbase.search

import android.content.Context
import com.amap.api.services.core.PoiItem
import com.amap.api.services.core.SuggestionCity
import com.amap.api.services.geocoder.GeocodeAddress
import com.amap.api.services.geocoder.GeocodeQuery
import com.amap.api.services.geocoder.GeocodeResult
import com.amap.api.services.poisearch.*
import com.amap.api.services.route.*
import com.amap.api.services.routepoisearch.RoutePOIItem
import com.amap.api.services.routepoisearch.RoutePOISearch
import com.amap.api.services.routepoisearch.RoutePOISearchQuery
import com.amap.api.services.routepoisearch.RoutePOISearchResult

class LatLng(
        val latitude: Double,
        val longitude: Double
)

class RoutePlanParam(
        /// 起点
        val from: LatLng,
        /// 终点
        val to: LatLng,
        /// 计算路径的模式，可选，默认为速度优先=0
        val mode: Int,
        /// 途经点，可选
        val passedByPoints: List<LatLng>?,
        /// 避让区域，可选，支持32个避让区域，每个区域最多可有16个顶点。如果是四边形则有4个坐标点，如果是五边形则有5个坐标点
        val avoidPolygons: List<List<LatLng>>?,
        /// 避让道路，只支持一条避让道路，避让区域和避让道路同时设置，只有避让道路生效
        val avoidRoad: String?
) {
    override fun toString(): String {
        return "RoutePlanParam(from=$from, to=$to, mode=$mode, passedByPoints=$passedByPoints, avoidPolygons=$avoidPolygons, avoidRoad='$avoidRoad')"
    }
}

//region UnifiedDriveRouteResult
internal class UnifiedDriveRouteResult(driveRouteResult: DriveRouteResult) {
    /// 起始
    val startPos: LatLng = driveRouteResult.startPos.toLatLng()

    /// 终点
    val targetPos: LatLng = driveRouteResult.targetPos.toLatLng()

    /// 打的费用
    val taxiCost: Double = driveRouteResult.taxiCost.toDouble()

    /// 路段
    val paths: List<UnifiedDrivePath> = driveRouteResult.paths.map { UnifiedDrivePath(it) }
}


class UnifiedDrivePath(drivePath: DrivePath) {
    val strategy: String = drivePath.strategy
    val tolls: Double = drivePath.tolls.toDouble()
    val tollDistance: Double = drivePath.tollDistance.toDouble()
    val totalTrafficlights: Int = drivePath.totalTrafficlights
    val steps: List<UnifiedDriveStep> = drivePath.steps.map { UnifiedDriveStep(it) }
    val restriction: Int = drivePath.restriction
}

class UnifiedDriveStep(driveStep: DriveStep) {
    val instruction: String = driveStep.instruction
    val orientation: String = driveStep.orientation
    val road: String = driveStep.road
    val distance: Double = driveStep.distance.toDouble()
    val tolls: Double = driveStep.tolls.toDouble()
    val tollDistance: Double = driveStep.tollDistance.toDouble()
    val tollRoad: String = driveStep.tollRoad
    val duration: Double = driveStep.duration.toDouble()
    val polyline: List<LatLng> = driveStep.polyline.map { it.toLatLng() }
    val action: String = driveStep.action
    val assistantAction: String = driveStep.assistantAction
    val routeSearchCityList: List<UnifiedRouteSearchCity> = driveStep.routeSearchCityList.map { UnifiedRouteSearchCity(it) }
    val TMCs: List<UnifiedTMC> = driveStep.tmCs.map { UnifiedTMC(it) }
}

class UnifiedRouteSearchCity(routeSearchCity: RouteSearchCity) {
    val districts: List<UnifiedDistrict> = routeSearchCity.districts.map { UnifiedDistrict(it) }
}

class UnifiedTMC(tmc: TMC) {
    val distance: Int = tmc.distance
    val status: String = tmc.status
    val polyline: List<LatLng> = tmc.polyline.map { it.toLatLng() }
}

class UnifiedDistrict(district: District) {
    val districtName: String = district.districtName
    val districtAdcode: String = district.districtAdcode
}
//endregion

//region UnifiedGeocodeResult
class UnifiedGeocodeResult(geocodeResult: GeocodeResult) {
    val geocodeQuery: UnifiedGeocodeQuery = UnifiedGeocodeQuery(geocodeResult.geocodeQuery)
    val geocodeAddressList: List<UnifiedGeocodeAddress> = geocodeResult.geocodeAddressList.map { UnifiedGeocodeAddress(it) }
}

class UnifiedGeocodeAddress(geocodeAddress: GeocodeAddress) {
    val formatAddress: String = geocodeAddress.formatAddress
    val province: String = geocodeAddress.province
    val city: String = geocodeAddress.city
    val district: String = geocodeAddress.district
    val township: String = geocodeAddress.township
    val neighborhood: String = geocodeAddress.neighborhood
    val building: String = geocodeAddress.building
    val adcode: String = geocodeAddress.adcode
    val latLng: LatLng = geocodeAddress.latLonPoint.toLatLng()
    val level: String = geocodeAddress.level
}

class UnifiedGeocodeQuery(geocodeQuery: GeocodeQuery) {
    val locationName: String = geocodeQuery.locationName
    val city: String = geocodeQuery.city
}
//endregion

//region UnifiedPoiItem
class UnifiedPoiItem(poiItem: PoiItem) {
    val businessArea: String? = poiItem.businessArea
    val adName: String? = poiItem.adName
    val cityName: String? = poiItem.cityName
    val provinceName: String? = poiItem.provinceName
    val typeDes: String? = poiItem.typeDes
    val tel: String? = poiItem.tel
    val adCode: String? = poiItem.adCode
    val poiId: String? = poiItem.poiId
    val distance: Int? = poiItem.distance
    val title: String? = poiItem.title
    val snippet: String? = poiItem.snippet
    val latLonPoint: LatLng? = poiItem.latLonPoint?.toLatLng()
    val cityCode: String? = poiItem.cityCode
    val enter: LatLng? = poiItem.enter?.toLatLng()
    val exit: LatLng? = poiItem.exit?.toLatLng()
    val website: String? = poiItem.website
    val postcode: String? = poiItem.postcode
    val email: String? = poiItem.email
    val direction: String? = poiItem.direction
    val isIndoorMap: Boolean? = poiItem.isIndoorMap
    val provinceCode: String? = poiItem.provinceCode
    val parkingType: String? = poiItem.parkingType
    val subPois: List<UnifiedSubPoiItem> = poiItem.subPois?.map { UnifiedSubPoiItem(it) }
            ?: listOf()
    val indoorData: UnifiedIndoorData? = if (poiItem.indoorData != null) UnifiedIndoorData(poiItem.indoorData) else null
    val photos: List<UnifiedPhoto> = poiItem.photos?.map { UnifiedPhoto(it) } ?: listOf()
    val poiExtension: UnifiedPoiItemExtension? = if (poiItem.poiExtension != null) UnifiedPoiItemExtension(poiItem.poiExtension) else null
    val typeCode: String? = poiItem.typeCode
    val shopID: String? = poiItem.shopID
}

class UnifiedPoiResult(poiResult: PoiResult) {
    val pageCount: Int? = poiResult.pageCount
    /**
     * 暂未支持
     */
    val query: PoiSearch.Query = poiResult.query
    private val bound: UnifiedSearchBound? = if (poiResult.bound != null) UnifiedSearchBound(poiResult.bound) else null
    private val pois: List<UnifiedPoiItem> = poiResult.pois?.map { UnifiedPoiItem(it) } ?: listOf()
    private val searchSuggestionKeywords: List<String> = poiResult.searchSuggestionKeywords
            ?: listOf()
    private val searchSuggestionCitys: List<UnifiedSuggestionCity> = poiResult.searchSuggestionCitys?.map { UnifiedSuggestionCity(it) }
            ?: listOf()
}

class UnifiedSubPoiItem(subPoiItem: SubPoiItem) {
    val poiId: String? = subPoiItem.poiId
    val title: String? = subPoiItem.title
    val subName: String? = subPoiItem.subName
    val distance: Int? = subPoiItem.distance
    val latLonPoint: LatLng? = subPoiItem.latLonPoint?.toLatLng()
    val snippet: String? = subPoiItem.snippet
    val subTypeDes: String? = subPoiItem.subTypeDes
}

class UnifiedIndoorData(indoorData: IndoorData) {
    val poiId: String? = indoorData.poiId
    val floor: Int? = indoorData.floor
    val floorName: String? = indoorData.floorName
}

class UnifiedPhoto(photo: Photo) {
    val title: String? = photo.title
    val url: String? = photo.url
}

class UnifiedPoiItemExtension(poiItemExtension: PoiItemExtension) {
    val opentime: String? = poiItemExtension.opentime
    val rating: String? = poiItemExtension.getmRating()
}

class UnifiedSuggestionCity(suggestionCity: SuggestionCity) {
    val cityName: String? = suggestionCity.cityName
    val cityCode: String? = suggestionCity.cityCode
    val adCode: String? = suggestionCity.adCode
    val suggestionNum: Int? = suggestionCity.suggestionNum
}
//endregion

//region UnifiedPoiSearchQuery
class UnifiedPoiSearchQuery(
        /// 查询字符串，多个关键字用“|”分割
        private val query: String,

        /// 待查询建筑物的标识
        private val building: String?,

        /// 待查分类组合
        private val category: String?,

        /// 待查城市（地区）的电话区号
        private val city: String,

        /// 设置查询的是第几页，从0开始
        private val pageNum: Int,

        /// 设置的查询页面的结果数目
        private val pageSize: Int,

        /// 是否严格按照设定城市搜索
        private val cityLimit: Boolean,

        /// 是否按照父子关系展示POI
        private val requireSubPois: Boolean,

        /// 是否按照距离排序
        private val distanceSort: Boolean,

        /// 设置的经纬度
        private val location: LatLng?,

        /// 搜索边界, 周边检索使用
        private val searchBound: UnifiedSearchBound?
) {
    /**
     * 关键字搜索
     */
    fun toPoiSearch(context: Context): PoiSearch {
        return PoiSearch(context, PoiSearch.Query(query, category, city).apply {
            building = this@UnifiedPoiSearchQuery.building
            pageNum = this@UnifiedPoiSearchQuery.pageNum
            pageSize = this@UnifiedPoiSearchQuery.pageSize
            cityLimit = this@UnifiedPoiSearchQuery.cityLimit
            requireSubPois(requireSubPois)
            isDistanceSort = this@UnifiedPoiSearchQuery.distanceSort
            location = this@UnifiedPoiSearchQuery.location?.toLatLonPoint()
        })
    }

    /**
     * 周边检索
     */
    fun toPoiSearchBound(context: Context): PoiSearch {
        return PoiSearch(context, PoiSearch.Query(query, category, city).apply {
            building = this@UnifiedPoiSearchQuery.building
            pageNum = this@UnifiedPoiSearchQuery.pageNum
            pageSize = this@UnifiedPoiSearchQuery.pageSize
            cityLimit = this@UnifiedPoiSearchQuery.cityLimit
            requireSubPois(requireSubPois)
            isDistanceSort = this@UnifiedPoiSearchQuery.distanceSort
            location = this@UnifiedPoiSearchQuery.location?.toLatLonPoint()
        }).apply { bound = searchBound?.toSearchBoundCenterRange() }
    }

    /**
     * 多边形内检索
     */
    fun toPoiSearchPolygon(context: Context): PoiSearch {
        return PoiSearch(context, PoiSearch.Query(query, category, city).apply {
            building = this@UnifiedPoiSearchQuery.building
            pageNum = this@UnifiedPoiSearchQuery.pageNum
            pageSize = this@UnifiedPoiSearchQuery.pageSize
            cityLimit = this@UnifiedPoiSearchQuery.cityLimit
            requireSubPois(requireSubPois)
            isDistanceSort = this@UnifiedPoiSearchQuery.distanceSort
            location = this@UnifiedPoiSearchQuery.location?.toLatLonPoint()
        }).apply { bound = searchBound?.toSearchBoundPolygon() }
    }
}
//endregion

//region UnifiedRoutePoiSearchQuery
class UnifiedRoutePoiSearchQuery(
        private val from: LatLng,
        private val to: LatLng,
        private val mode: Int,
        private val searchType: Int,
        private val range: Int,
        private val polylines: List<LatLng>
) {
    fun toRoutePoiSearchLine(context: Context): RoutePOISearch {
        return RoutePOISearch(
                context,
                RoutePOISearchQuery(
                        from.toLatLonPoint(),
                        to.toLatLonPoint(),
                        mode,
                        when (searchType) {
                            0 -> RoutePOISearch.RoutePOISearchType.TypeGasStation
                            1 -> RoutePOISearch.RoutePOISearchType.TypeMaintenanceStation
                            2 -> RoutePOISearch.RoutePOISearchType.TypeATM
                            3 -> RoutePOISearch.RoutePOISearchType.TypeToilet
                            4 -> RoutePOISearch.RoutePOISearchType.TypeFillingStation
                            5 -> RoutePOISearch.RoutePOISearchType.TypeServiceArea
                            else -> throw IllegalArgumentException()
                        },
                        range
                )
        )
    }

    fun toRoutePoiSearchPolygon(context: Context): RoutePOISearch {
        return RoutePOISearch(
                context,
                RoutePOISearchQuery(
                        polylines.map { it.toLatLonPoint() },
                        when (searchType) {
                            0 -> RoutePOISearch.RoutePOISearchType.TypeGasStation
                            1 -> RoutePOISearch.RoutePOISearchType.TypeMaintenanceStation
                            2 -> RoutePOISearch.RoutePOISearchType.TypeATM
                            3 -> RoutePOISearch.RoutePOISearchType.TypeToilet
                            4 -> RoutePOISearch.RoutePOISearchType.TypeFillingStation
                            5 -> RoutePOISearch.RoutePOISearchType.TypeServiceArea
                            else -> throw IllegalArgumentException()
                        },
                        range
                )
        )
    }
}
//endregion

//region UnifiedRoutePoiSearchResult
class UnifiedRoutePoiSearchResult(routePOISearchResult: RoutePOISearchResult) {
    private val routePoiList: List<UnifiedRoutePOIItem> = routePOISearchResult.routePois?.map { UnifiedRoutePOIItem(it) }
            ?: listOf()
    private val query: RoutePOISearchQuery = routePOISearchResult.query
}

private class UnifiedRoutePOIItem(poiItem: RoutePOIItem) {
    val id: String = poiItem.id
    val title: String = poiItem.title
    val point: LatLng? = poiItem.point?.toLatLng()
    val distance: Float = poiItem.distance
    val duration: Float = poiItem.duration
}
//endregion

class UnifiedSearchBound(searchBound: PoiSearch.SearchBound) {
    val lowerLeft: LatLng? = searchBound.lowerLeft?.toLatLng()
    val upperRight: LatLng? = searchBound.upperRight?.toLatLng()
    private val center: LatLng? = searchBound.center?.toLatLng()
    private val range: Int? = searchBound.range
    val shape: String? = searchBound.shape
    val isDistanceSort: Boolean? = searchBound.isDistanceSort
    private val polyGonList: List<LatLng> = searchBound.polyGonList?.map { it.toLatLng() } ?: listOf()

    /**
     * 中心点 + 半径范围
     */
    fun toSearchBoundCenterRange(): PoiSearch.SearchBound {
        return PoiSearch.SearchBound(center?.toLatLonPoint(), range!!)
    }

    /**
     * 多边形搜索
     */
    fun toSearchBoundPolygon(): PoiSearch.SearchBound {
        return PoiSearch.SearchBound(polyGonList.map { it.toLatLonPoint() })
    }
}